aten/src/ATen/native/xnnpack/Convolution.cpp

```diff
@@ -183,67 +137,111 @@ ContextConv2D create(
-      Operator(convolution_op),
-      {weight_nhwc.sizes()[0], weight_nhwc.sizes()[1],
-          weight_nhwc.sizes()[2], weight_nhwc.sizes()[3]},
-      {padding_expanded[0], padding_expanded[1]},
-      {stride_expanded[0], stride_expanded[1]},
-      {dilation_expanded[0], dilation_expanded[1]}
-  };
+  return ContextConv2D{Operator(convolution_op),
+                       {weight_nhwc.sizes()[0],
+                        weight_nhwc.sizes()[1],
+                        weight_nhwc.sizes()[2],
+                        weight_nhwc.sizes()[3]},
+                       {padding_expanded[0], padding_expanded[1]},
+                       {stride_expanded[0], stride_expanded[1]},
+                       {dilation_expanded[0], dilation_expanded[1]}};
 }
 
-c10::intrusive_ptr<xnnpack::XNNPackConv2dOpContext> Conv2dPrePack::operator()(
-  Tensor weight,
-  c10::optional<Tensor> bias,
-  std::vector<int64_t> stride,
-  std::vector<int64_t> padding,
-  std::vector<int64_t> dilation,
-  int64_t groups
-  ) {
-    return xnnpack::XNNPackConv2dOpContext::create_context(
-        std::move(weight),
-        std::move(bias),
-        std::move(padding),
-        std::move(stride),
-        std::move(dilation),
-        groups,
-        {},
-        {});
+Tensor run(const ContextConv2D& context, const Tensor& input) {
+  using namespace internal;
+
+  const Tensor input_nhwc = input.contiguous(MemoryFormat::ChannelsLast);
+  const Tensor padded_input_nhwc = allocate_padded_if_needed(input_nhwc);
+
+  TORCH_CHECK(
+      usable(padded_input_nhwc),
+      "XNNPACK Convolution not usable! "
+      "Reason: The provided input tensor is either invalid or unsupported by XNNPACK.");
+
+  Tensor output = empty_with_tail_padding(
+      conv_output_size(
+          padded_input_nhwc.sizes(),
+          context.weight_size_,
+          context.padding_,
+          context.stride_,
+          context.dilation_),
+      padded_input_nhwc.options().dtype(),
+      MemoryFormat::ChannelsLast);
+
+  const xnn_status setup_status = xnn_setup_convolution2d_nhwc_f32(
+      context.op.get(), // operator
+      padded_input_nhwc.size(Layout::Activation4D::batch), // batch_size
+      padded_input_nhwc.size(Layout::Activation4D::height), // input_height
+      padded_input_nhwc.size(Layout::Activation4D::width), // input_width
+      padded_input_nhwc.data_ptr<float>(), // input
+      output.data_ptr<float>(), // output
+      caffe2::xnnpack_threadpool()); // threadpool
+
+  TORCH_CHECK(
+      xnn_status_success == setup_status,
+      "xnn_setup_convolution2d_nhwc_f32 failed!");
+
+  const xnn_status run_status = xnn_run_operator(
+      context.op.get(), // operator
+      caffe2::xnnpack_threadpool()); // threadpool
+
+  TORCH_INTERNAL_ASSERT(
+      xnn_status_success == run_status, "xnn_run_operator failed!");
+
+  return output.contiguous(input.suggest_memory_format());
+}
+
+c10::intrusive_ptr<xnnpack::Conv2dOpContext> Conv2dPrePack::operator()(
```

Comment:
nit: you don't need this trampoline - you can register `xnnpack::XNNPackConv2dOpContext::create_context` directly with RegisterOps (it doesn't need to be a functor)

Reply:
I tried that in the first diff itself and something I ran into that would not let me compile, cannot remember what. I will try again and see if there is a way around else report the error back here.